### PR TITLE
extracting attributes for grover model from batch graph

### PR DESCRIPTION
**Pull request by gusty1g**
## Description

Related to #3296 

This pull request proposes a new approach to extract required attributes for a grover model from a batched molecular graph.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/feat/graph_data.py
Comment (rbharath) [inline deepchem/feat/graph_data.py:339]: Can you remove TODO before merge?

File: deepchem/feat/molecule_featurizers/grover_featurizer.py
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/grover_featurizer.py:125]: Why are we deleting these?

File: deepchem/feat/molecule_featurizers/grover_featurizer.py
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/grover_featurizer.py:156]: Why are we deleting these?

File: deepchem/models/torch_models/grover.py
Comment (rbharath) [inline deepchem/models/torch_models/grover.py:26]: To confirm, we are removing embedding_output_type since it is not described in the Grover paper?

File: deepchem/utils/grover.py
Comment (rbharath) [inline deepchem/utils/grover.py:14]: Can you add a `Parameters` section here documenting the types of the argument?

File: deepchem/utils/grover.py
Comment (rbharath) [inline deepchem/utils/grover.py:35]: Can you add a `Parameters` section here documenting the types of the arguments?

File: deepchem/utils/grover.py
Comment (rbharath) [inline deepchem/utils/grover.py:56]: Can you add a `Parameters` section here documenting the types of the argument?

File: deepchem/utils/grover.py
Comment (rbharath) [inline deepchem/utils/grover.py:74]: Can you add a `Parameters` section here documenting the types of the argument?

File: deepchem/utils/grover.py
Comment (rbharath) [inline deepchem/utils/grover.py:103]: typo: Parameters (missing "s")

**Review by rbharath — COMMENTED** [review]
Design in general looks good here. Some minor comments and questions below that should be straightforward to address

File: deepchem/feat/molecule_featurizers/grover_featurizer.py
Comment (gusty1g) [reply deepchem/feat/molecule_featurizers/grover_featurizer.py:125]: The idea behind these changes is to extract `a2b` and other attributes from the graph using edge index. This can be done by the newly added `extract_grover_attributes` method. So, the attributes extracted from featurizer are not needed.

File: deepchem/feat/molecule_featurizers/grover_featurizer.py
Comment (gusty1g) [reply deepchem/feat/molecule_featurizers/grover_featurizer.py:156]: These attributes can be extracted from `extract_grover_attributes` method.

File: deepchem/models/torch_models/grover.py
Comment (gusty1g) [reply deepchem/models/torch_models/grover.py:26]: yes

File: deepchem/utils/grover.py
Comment (gusty1g) [reply deepchem/utils/grover.py:14]: added

File: deepchem/utils/grover.py
Comment (gusty1g) [reply deepchem/utils/grover.py:35]: added

File: deepchem/utils/grover.py
Comment (gusty1g) [reply deepchem/utils/grover.py:56]: added

File: deepchem/utils/grover.py
Comment (gusty1g) [reply deepchem/utils/grover.py:74]: added

File: deepchem/utils/grover.py
Comment (gusty1g) [reply deepchem/utils/grover.py:103]: fixed type

File: deepchem/feat/graph_data.py
Comment (gusty1g) [reply deepchem/feat/graph_data.py:339]: I have removed TODO but a question I have is should we rename `graph_index` as `node_index` because I think `node_index` is more descriptive. The variable here stores a mapping for every node and the graph to which the node belongs to.

**Review by rbharath — APPROVED** [review]
LGTM, feel free to merge once CI is clear

**Comment by rbharath** [comment]
@gusty1g Looks like some mypy errors which should be fixed before merge

**Pull request merged**
